[Wasm-GC] Fix regression on armv7 in structs.js test

takikawa opened the pull request:
#### 97d8c872ff5e463bc03701fcca5c83000b0e3bf4
<pre>
[Wasm-GC] Fix regression on armv7 in structs.js test
<a href="https://bugs.webkit.org/show_bug.cgi?id=244352">https://bugs.webkit.org/show_bug.cgi?id=244352</a>

Reviewed by Justin Michaud.

Fix a write to the call frame header&apos;s callee slot that didn&apos;t account
for the cell tag on armv7. Also refactors some related code to use the
same helper for this.

* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::emitPutCellToCallFrameHeader):
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::handleBadGCTypeIndexUse):
(JSC::Wasm::wasmToJS):

Canonical link: <a href="https://commits.webkit.org/253906@main">https://commits.webkit.org/253906@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a2b480f4c7a25c25a644cb31de6c58373a22ada

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87392 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31480 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18201 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96621 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149968 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29844 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26037 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79509 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91392 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93009 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24111 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74182 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23950 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79082 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79309 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66967 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79241 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27559 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13142 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72922 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27510 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14158 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25969 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2744 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29198 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37019 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75743 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29128 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33436 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16784 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
```diff
@@ -61,7 +61,8 @@ static Expected<MacroAssemblerCodeRef<WasmEntryPtrTag>, BindingFailure> handleBa
     jit.loadPtr(CCallHelpers::Address(GPRInfo::argumentGPR0, JSWebAssemblyInstance::offsetOfModule()), GPRInfo::argumentGPR1);
     jit.prepareCallOperation(vm);
-    jit.emitPutToCallFrameHeader(GPRInfo::argumentGPR1, CallFrameSlot::callee);
+    CCallHelpers::Address calleeSlot { GPRInfo::callFrameRegister, CallFrameSlot::callee  * sizeof(Register) };
```
Are all uses of jit.emitPutToCallFrameHeader wrong on ARMv7 then?

justinmichaud replied:
Should we change that function instead?

takikawa replied:
I don't think this applies to all instances of that function call, because it's used on different slots in the call frame. It's specifically the callee slot that's the issue here, while the function is also used on `CallFrameSlot::codeBlock` instead for example.

There might be ~2 other calls that would be good to check on though.

justinmichaud replied:
Maybe we can make a helper, putCellToFrame?

takikawa replied:
I added a helper in the latest version, and refactored one other case in `WasmToJS.cpp` to use the helper (there is another bit of code modifying the cell tag too, but it's touching the callee frame so the helper doesn't apply).

justinmichaud reviewed (approved):
r=me, up to you if this shoudl be a helper / if other sites are affected

takikawa committed:
[Wasm-GC] Fix regression on armv7 in structs.js test
https://bugs.webkit.org/show_bug.cgi?id=244352

Reviewed by Justin Michaud.

Fix a write to the call frame header's callee slot that didn't account
for the cell tag on armv7. Also refactors some related code to use the
same helper for this.

* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::emitPutCellToCallFrameHeader):
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::handleBadGCTypeIndexUse):
(JSC::Wasm::wasmToJS):

Canonical link: https://commits.webkit.org/253906@main
